contracts/WooRebateManager.sol

```diff
@@ -0,0 +1,142 @@
+
+import '@openzeppelin/contracts/utils/ReentrancyGuard.sol';
+import '@openzeppelin/contracts/math/SafeMath.sol';
+import '@openzeppelin/contracts/token/ERC20/SafeERC20.sol';
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import '@openzeppelin/contracts/token/ERC20/ERC20.sol';
+import '@uniswap/lib/contracts/libraries/TransferHelper.sol';
+
+contract WooRebateManager is InitializableOwnable, IWooRebateManager {
+    using SafeMath for uint256;
+    using DecimalMath for uint256;
+    using SafeERC20 for IERC20;
+
+    // Note: this is the percent rate of the total swap fee (not the swap volume)
+    // decimal: 18; 1e16 = 1%, 1e15 = 0.1%, 1e16 = 0.01%
+    //
+    // e.g. suppose:
+    //   rebateRate = 1e17 (10%), so the rebate amount is total_swap_fee * 10%.
+    mapping(address => uint256) public override rebateRate;
+
+    // pending rebate amount in quote token
+    mapping(address => uint256) public pendingRebate;
+
+    IWooPP private wooPP;
+
+    address public immutable quoteToken; // USDT
+    address public immutable rewardToken; // WOO
+
+    constructor(address newQuoteToken, address newRewardToken) public {
+        require(newQuoteToken != address(0), 'WooRebateManager: INVALID_QUOTE');
+        require(newRewardToken != address(0), 'WooRebateManager: INVALID_RAWARD_TOKEN');
+        initOwner(msg.sender);
+        quoteToken = newQuoteToken;
+        rewardToken = newRewardToken;
+    }
+
+    function addRebate(address brokerAddr, uint256 amountInUSD) external override {
+        if (brokerAddr == address(0)) {
+            return;
+        }
+
+        uint256 balanceBefore = IERC20(quoteToken).balanceOf(address(this));
+        TransferHelper.safeTransferFrom(quoteToken, msg.sender, address(this), amountInUSD);
+        uint256 balanceAfter = IERC20(quoteToken).balanceOf(address(this));
+        require(balanceAfter.sub(balanceBefore) >= amountInUSD, 'RebateManager: amount < balance delta');
+
+        pendingRebate[brokerAddr] = amountInUSD.add(pendingRebate[brokerAddr]);
+    }
+
+    function pendingRebateInUSDT(address brokerAddr) external view override returns (uint256) {
+        require(brokerAddr != address(0), 'WooRebateManager: zero_brokerAddr');
+        return pendingRebate[brokerAddr];
+    }
+
+    function pendingRebateInWoo(address brokerAddr) external view override returns (uint256) {
+        require(brokerAddr != address(0), 'WooRebateManager: zero_brokerAddr');
+        return wooPP.querySellQuote(rewardToken, pendingRebate[brokerAddr]);
+    }
+
+    function claimRebate() external override {
+        require(pendingRebate[msg.sender] > 0, 'WooRebateManager: NO_pending_rebate');
+
+        uint256 quoteAmount = pendingRebate[msg.sender];
+        uint256 balanceBefore = IERC20(rewardToken).balanceOf(address(this));
+        TransferHelper.safeApprove(quoteToken, address(wooPP), quoteAmount);
+        uint256 wooAmount = wooPP.sellQuote(rewardToken, quoteAmount, 0, address(this), address(0));
```

Comment:
third parameters in `sellQuote` should not be `0`, prevent from being `sandwich attack`, should be query from `querySellQuote` first